vac/raw/decentralized-messaging-ethereum.md

```diff
@@ -0,0 +1,737 @@
+even if that member is currently offline.
+
+##### 2SM-Send
+This function takes a state `sigma` and a plaintext `m` as inputs, and 
+returns a new state `sigma’` and a ciphertext `c`.
+
+##### 2SM-Receive
+This function takes a state `sigma` and a ciphertext `c`, and 
+returns a new state `sigma’` and a plaintext `m`.
+
+#### 2SM Syntax
+
+The variable `sigma` denotes the state consisting in the variables below:
+```
+sigma.mySks[0] = sk
+sigma.nextIndex = 1 
+sigma.receivedSk = empty_string
+sigma.otherPk = pk`<br> 
+sigma.otherPksender = “other”
+sigma.otherPkIndex = 0
+```
+
+#### 2SM-Init
+On input a key pair `(sk, pk)`, this functions otuputs a state `sigma`.
+
+#### 2SM-Send
+This function encrypts the message `m` using `sigma.otherPk`, 
+which represents the other party’s current public key. 
+This key is determined based on the last public key generated for the other party or 
+the last public key received from the other party, 
+whichever is more recent. 
+`sigma.otherPkSender` is set to `me` in the former case and `other` in the latter case. 
+
+Metadata including `otherPkSender` and 
+`otherPkIndex` are included in the message to indicate which of the recipient’s public keys is being utilized.
+
+Additionally, this function generates a new key pair for the local user, 
+storing the secret key in `sigma.mySks` and sending the public key. 
+Similarly, it generates a new key pair for the other party, 
+sending the secret key (encrypted) and storing the public key in `sigma.otherPk`.
+
+```
+sigma.mySks[sigma.nextIndex], myNewPk) = PKE-Gen()
+(otherNewSk, otherNewPk) = PKE-Gen()
+plaintext = (m, otherNewSk, sigma`.nextIndex, myNewPk)
+msg = (PKE-Enc(sigma.otherPk, plaintext), sigma.otherPkSender, sigma.otherPkIndex)
+sigma.nextIndex++
+(sigma.otherPk, sigma.otherPkSender, sigma.otherPkIndex) = (otherNewPk, "me", empty_string)
+return (sigma`, msg)
+```
+
+#### 2SM-Receive
+
+This function utilizes the metadata of the message `c` to determine which secret key to utilize for decryption, 
+assigning it to `sk`. 
+If the secret key corresponds to one generated by ourselves, 
+that secret key along with all keys with lower index are deleted. 
+This deletion is indicated by `sigma.mySks[≤ keyIndex] = empty_string`. 
+Subsequently, the new public and secret keys contained in the message are stored.
+
+```
+(ciphertext, keySender, keyIndex) = c
+if keySender = "other" then 
+sk = sigma.mySks[keyIndex] 
+sigma.mySks[≤ keyIndex] = empty_string
+else sk = sigma.receivedSk
+(m, sigma.receivedSk, sigma.otherPkIndex, sigma.otherPk) = PKE-Dec(sk, ciphertext)
+sigma.otherPkSender = "other"
+return (sigma, m)
+```
+### PKE Syntax
+
+The required PKE that MUST be used is ElGamal with a 2048-bit modulus `p`. 
+
+#### Parameters
+
+The following parameters must be used:
+
+```
+p = 308920927247127345254346920820166145569
+g = 2
+```
+
+#### PKE-KGen
+
+Each user `u` MUST do the following:
+
+```
+PKE-KGen():
+a = randint(1, p-2)
```

Comment:
Little suggestion. Maybe it's better to choose a in interval [2, p-2] for all further computation regarding random value it ElGamal. I mean DLP is not a problem if our secret is 1 :) 

That's true that in original white paper interval is [1;p-2]. But in later work regarding prevent attack I saw that interval [2;p-2] is better